### ISSUE TITLE
TPUStrategy breaks on subclassed keras models

### ISSUE THREAD

**Issue by kazimuth** [tensorflow/tensorflow#33657]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab TPU
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d38 2.0.0
- Python version: 3.7

**Code to reproduce the issue**

```python
%tensorflow_version 2.x

# setup

import tensorflow as tf
import os

tpu_address = 'grpc://' + os.environ['COLAB_TPU_ADDR']
cluster_resolver = tf.distribute.cluster_resolver.TPUClusterResolver(tpu=tpu_address)
tf.config.experimental_connect_to_cluster(cluster_resolver)
tf.tpu.experimental.initialize_tpu_system(cluster_resolver)
STRATEGY = tf.compat.v1.distribute.experimental.TPUStrategy(cluster_resolver)

# define a subclassed keras model
class Simple(tf.keras.Model):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.dense = tf.keras.layers.Dense(10)

    def call(self, x, training=None):
        return self.dense(x, training=training)

# a fake dataset
data = tf.data.Dataset.from_tensor_slices((
    tf.random.uniform((10, 10)),
    tf.random.uniform((10, 10))
)).batch(10).repeat()

# outside the TPUStrategy scope, the model works fine...
model = Simple()
model.compile(loss='categorical_crossentropy')
model.evaluate(data, steps=1)
# ==> 1/1 [==============================] - 0s 62ms/step - loss: 40.6148

# but if we try and use the model in the scope...
with STRATEGY.scope():
    model = Simple()
    model.compile(loss='categorical_crossentropy')

model.evaluate(data, steps=1)

# we get an error:
# 
# ValueError                                Traceback (most recent call last)
# 
# <ipython-input-17-56ee0b4381e4> in <module>()
#      29     model = Nested(model)
#      30     model.compile(loss='categorical_crossentropy')
# ---> 31 model.evaluate(data, steps=1)
# 
# ... 12 frames ...
# 
# /tensorflow-2.0.0/python3.6/tensorflow_core/python/framework/func_graph.py in wrapper(*args, **kwargs)
#     903           except Exception as e:  # pylint:disable=broad-except
#     904             if hasattr(e, "ag_error_metadata"):
# --> 905               raise e.ag_error_metadata.to_exception(e)
#     906             else:
#     907               raise
# 
# ValueError: in converted code:
#     relative to /tensorflow-2.0.0/python3.6/tensorflow_core/python/keras:
# 
#     distribute/distributed_training_utils.py:862 distributed_function  *
#         x, y, sample_weights = input_fn()
#     engine/training_arrays.py:516 get_distributed_inputs
#         model, inputs, targets, sample_weights, mode)
#     distribute/distributed_training_utils.py:638 _prepare_feed_values
#         inputs, targets, sample_weights = _get_input_from_iterator(inputs, model)
#     distribute/distributed_training_utils.py:616 _get_input_from_iterator
#         x, y, sample_weights = next_element
# 
#     ValueError: not enough values to unpack (expected 3, got 2)
```

**Expected behavior**

This should not throw any error.

**Comment by kazimuth** [comment]
Update: found a simpler repro.

**Comment by gadagashwini** [comment]
Issue is replicating on colab with TF 2.0. Please take a look at [gist](https://colab.sandbox.google.com/gist/gadagashwini/4060275d458dc7547a3ab4a6a670f126/untitled225.ipynb). Thanks!

**Comment by frankchn** [comment]
TPUs are not officially supported in TF2.0. Does the code work on 1.15 or a nightly version?

**Comment by jvishnuvardhan** [comment]
@frankchn TF1.15.0 throws a different error as shown below. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/13ae47a9d0131514db9e760eec777e9f/untitled225.ipynb). 

Error trace
```
INFO:tensorflow:Initializing the TPU system: 10.104.18.114:8470
INFO:tensorflow:Finished initializing TPU system.
INFO:tensorflow:Querying Tensorflow master (grpc://10.104.18.114:8470) for TPU system metadata.
INFO:tensorflow:Found TPU system:
INFO:tensorflow:*** Num TPU Cores: 8
INFO:tensorflow:*** Num TPU Workers: 1
INFO:tensorflow:*** Num TPU Cores Per Worker: 8
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:CPU:0, CPU, -1, 12994194352412773709)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:0, TPU, 17179869184, 3779494438081468292)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:1, TPU, 17179869184, 7546761505036203465)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:2, TPU, 17179869184, 15889901405504100012)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:3, TPU, 17179869184, 16304530786141444431)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:4, TPU, 17179869184, 15091146302727302848)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:5, TPU, 17179869184, 4279064904853472805)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:6, TPU, 17179869184, 13012250068625670504)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:7, TPU, 17179869184, 15180363532189083818)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU_SYSTEM:0, TPU_SYSTEM, 8589934592, 1141987253231946822)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:XLA_CPU:0, XLA_CPU, 17179869184, 2127150505038211523)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-b944ab9ba609> in <module>()
     34     model.compile(loss='categorical_crossentropy')
     35 
---> 36 model.evaluate(data, steps=1)

9 frames
/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/api.py in wrapper(*args, **kwargs)
    235       except Exception as e:  # pylint:disable=broad-except
    236         if hasattr(e, 'ag_error_metadata'):
--> 237           raise e.ag_error_metadata.to_exception(e)
    238         else:
    239           raise

ValueError: in converted code:
    relative to /usr/local/lib/python3.6/dist-packages/tensorflow_core/python:

    keras/engine/training_distributed.py:333 _test_step_fn
        _build_model, args=(model, mode, inputs, targets)))
    distribute/distribute_lib.py:1940 merge_call
        return self._merge_call(merge_fn, args, kwargs)
    distribute/distribute_lib.py:1947 _merge_call
        return merge_fn(self._strategy, *args, **kwargs)
    keras/engine/training_distributed.py:58 _build_model
        targets)
    keras/distribute/distributed_training_utils.py:783 _build_distributed_network
        args=(model, mode, inputs, targets))
    distribute/distribute_lib.py:1810 call_for_each_replica
        return self._call_for_each_replica(fn, args, kwargs)
    distribute/tpu_strategy.py:372 _call_for_each_replica
        return fn(*args, **kwargs)
    keras/distribute/distributed_training_utils.py:743 _build_network_on_replica
        model, input_tensors=inputs, layer_fn=models.share_weights)
    keras/models.py:166 _clone_functional_model
        raise ValueError('Expected `model` argument '

    ValueError: Expected `model` argument to be a functional `Model` instance, but got a subclass model instead.
```

**Comment by frankchn** [comment]
@rxsang can you help take a look? I am OOO today

**Comment by kazimuth** [comment]
Side note, this can be worked around easily enough w/ custom training loops.

**Comment by amahendrakar** [comment]
@kazimuth,
The issue seems to be resolved with TensorFlow v2.4. 

I was able to run the code without any issues, please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/8cb7247ff498cb58347d0b161a174fe3/33657.ipynb). Thanks!
